src/utils/queue_builder.rs

```diff
@@ -1,71 +1,41 @@
+use log::{debug, error, warn};
 use std::panic::{catch_unwind, AssertUnwindSafe};
 use std::sync::Arc;
 use std::time::Duration;
 use std::{fs, io, thread};
 
 pub(crate) const TEMPDIR_PREFIX: &str = "docsrs-docs";
 
-// TODO: change to `fn() -> Result<!, Error>` when never _finally_ stabilizes
 pub fn queue_builder(
     mut builder: RustwideBuilder,
     build_queue: Arc<BuildQueue>,
 ) -> Result<(), Error> {
-    /// Represents the current state of the builder thread.
-    enum BuilderState {
-        /// The builder thread has just started, and hasn't built any crates yet.
-        Fresh,
-        /// The builder has just seen an empty build queue.
-        EmptyQueue,
-        /// The builder has just seen the lock file.
-        Locked,
-        /// The builder has started (or just finished) building a crate.
-        QueueInProgress,
-    }
-
-    let mut status = BuilderState::Fresh;
-
     loop {
         if let Err(e) = remove_tempdirs() {
             report_error(&anyhow::anyhow!(e).context("failed to remove temporary directories"));
         }
 
-        if !matches!(status, BuilderState::QueueInProgress) {
-            thread::sleep(Duration::from_secs(60));
-        }
-
         // check lock file
-        if build_queue.is_locked() {
+        if build_queue.is_locked()? {
             warn!("Lock file exists, skipping building new crates");
-            status = BuilderState::Locked;
+            thread::sleep(Duration::from_secs(60));
             continue;
         }
 
-        // Only build crates if there are any to build
-        debug!("Checking build queue");
-        match build_queue.pending_count() {
-            Err(e) => {
-                report_error(&e.context("Failed to read the number of crates in the queue"));
-                continue;
-            }
-
-            Ok(0) => {
-                debug!("Queue is empty, going back to sleep");
-                status = BuilderState::EmptyQueue;
-                continue;
-            }
-
-            Ok(queue_count) => info!("Starting build with {} crates in queue", queue_count),
-        }
-
-        status = BuilderState::QueueInProgress;
-
         // If a panic occurs while building a crate, lock the queue until an admin has a chance to look at it.
+        debug!("Checking build queue");
         let res = catch_unwind(AssertUnwindSafe(|| {
-            if let Err(e) = build_queue.build_next_queue_package(&mut builder) {
-                report_error(&e.context("Failed to build crate from queue"));
+            match build_queue.build_next_queue_package(&mut builder) {
+                Ok(true) => {}
+                Ok(false) => {
+                    debug!("Queue is empty, going back to sleep");
+                    thread::sleep(Duration::from_secs(60));
+                }
+                Err(e) => {
+                    report_error(&e.context("Failed to build crate from queue"));
+                }
             }
         }));
 
```

Comment:
There's a comment below that says "If we panic here something is really truly wrong and trying to handle the error won't help". That no longer seems true to me, since the database could be down or whatever. But I don't know how to handle the error. Maybe we should treat it the same as the rest of the errors, exit the build loop and report it?

Reply:
For a moment I thought you meant exiting = `break`, but I think you mean `continue` like with the other errors. 

I don't know which cases can lead to panics in the build. I remember we still had some places sprinkled with `.unwrap` or `.expect` around database methods, where trying again could be valid. 

Checking this I also saw that the `?` when checking the queue lock should be removed to match the error-handling in other cases in that loop.

Reply:
I updated the logic , only reporting and `continue` also for panics

Reply:
IIUC, the failure case here if it's the crate causing the failure will just be that we retry building the top crate from the queue continuously every 60 seconds (because the panic bypasses incrementing the attempt counter); until someone manually removes that crate from the queue (which would have been the same resolution before). If it's something transient like the database, then we should eventually recover when it's working and we attempt to build the crate again.